x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -788,8 +789,8 @@ Collection<Object> createComponents(
         this.persistentTasksService.set(persistentTasksService);
 
         systemIndices.getMainIndexManager().addStateListener((oldState, newState) -> {
-            if (clusterService.state().nodes().isLocalNodeElectedMaster()) {
-                applyPendingSecurityMigrations(newState);
+            if (clusterService.state().nodes().isLocalNodeElectedMaster() && oldState != UNRECOVERED_STATE) {
```

Comment:
This piece is what was missing when I previously tried to get this working. 

The addition of this guarantees that we can now trust both the old and new state and therefore use it to determine if the security index was just created or not.

Reply:
I don't understand why this is needed. I think this amounts to skipping the very first notification for the security index state update (which has the oldState in its default value), but why?